lib/Sema/TypeCheckType.cpp

```diff
@@ -2662,19 +2662,26 @@ TypeResolver::resolveOpenedExistentialArchetype(
                     interfaceType);
 
     archetypeType = ErrorType::get(interfaceType->getASTContext());
-  } {
+  } else {
     // The constraint type is written with respect to the surrounding
     // generic environment.
     constraintType = GenericEnvironment::mapTypeIntoContext(
                resolution.getGenericSignature().getGenericEnvironment(),
                constraintType);
 
-    // The opened existential type is formed by mapping the interface type
-    // into a new opened generic environment.
-    archetypeType = OpenedArchetypeType::get(constraintType->getCanonicalType(),
-                                             interfaceType,
-                                             GenericSignature(),
-                                             attrs.getOpenedID());
+    if (auto *dmt = interfaceType->getAs<DependentMemberType>()) {
+      auto base = dmt->getBase();
+      auto baseArchetype =
+          OpenedArchetypeType::get(constraintType->getCanonicalType(), base,
+                                   GenericSignature(), attrs.getOpenedID());
+      archetypeType = DependentMemberType::get(baseArchetype, dmt->getName());
```

Comment:
I see two problems here:
- archetypeType is a DependentMemberType in one branch and an OpenedArchetypeType in another. That doesn't seem right?
- you're only unwrapping one level of dependent member type (you can have an opened archetype for `Self.A.B.C`, etc.) Perhaps you want to use TypeBase::getRootGenericParam()?

Reply:
> Perhaps you want to use TypeBase::getRootGenericParam()?
Thanks!  Changed it to call from `DependentMemberType::getBase` to `TypeBase::getRootGenericParam()` and unwrap all the levels of dependent member types.
> archetypeType is a DependentMemberType in one branch and an OpenedArchetypeType in another. That doesn't seem right?
Perhaps a different variable name is in order.  It seems only to need to be a `Type`.  On another branch it's ErrorType.  Could you say more specifically what's wrong?

Reply:
You're forming a DependentMemberType with an archetype base and storing this in a variable named `archetypeType`. This is not well-formed because the base of a DependentMemberType must be another DependentMemberType or a generic parameter.

I think you want to do the same `archetypeType = OpenedArchetypeType::get(...)` call in both branches to form the actual result that is returned, but in the case that the interface type is an archetype you also want to form the base archetype.